sdv/single_table/utils.py

```diff
@@ -27,24 +31,37 @@ def detect_discrete_columns(metadata, data):
             A list of discrete columns to be used with some of ``sdv`` synthesizers.
     """
     discrete_columns = []
-
     for column in data.columns:
-        if column in metadata.columns:
-            if metadata.columns[column]['sdtype'] not in ['numerical', 'datetime']:
-                discrete_columns.append(column)
+        # Numerical and datetime columns never get preprocessed into categorical ones
+        if column in metadata.columns and \
+                metadata.columns[column]['sdtype'] in ['numerical', 'datetime']:
+            continue
 
-        else:
-            column_data = data[column].dropna()
-            if set(column_data.unique()) == {0.0, 1.0}:
-                column_data = column_data.astype(bool)
+        column_data = data[column].dropna()
 
-            try:
-                dtype = column_data.infer_objects().dtype.kind
-                if dtype in ['O', 'b']:
-                    discrete_columns.append(column)
+        # Ignore columns with only nans and empty datasets
+        if column_data.empty:
+            continue
 
-            except Exception:
-                discrete_columns.append(column)
+        # Non-integer floats and integers with too many unique values cannot be categorical
+        try:
+            column_data = column_data.astype('float')
+            rounded_column_data = list(column_data.round())
+            column_data = list(column_data)
+            is_float = column_data != rounded_column_data
+            is_int = column_data == rounded_column_data
+            num_values = len(column_data)
+            num_categories = len(set(column_data))
+            threshold = max(10, num_values * .1)
```

Comment:
Why `max(10, num_values * .1)` here and not `num_values * .1`?

Reply:
The main reason not to treat it as categorical is because it's very slow for many categories, because CTGAN one hot encodes them. If you have less than 10 categories, that's not a problem, since you will have at most 10 extra columns, and your dataset is very small (< 100 rows). So I decided to cap it, if your data has less than 10 unique values, it will always be treated as categorical.

Reply:
I agree, thanks for the explanation